### ISSUE TITLE
Overlap removal

### ISSUE THREAD

**Issue by elalish** [elalish/manifold#289]
This will be a large feature; I'm not sure when I'll have time to tackle it, but I want to write down my thoughts since I now have a sketch of the algorithm. 

Removing overlaps would have two major uses: as the final stage of a manifold fixing algorithm for non-manifold input meshes, and also as a cleanup after `Warp()` that would take care of produced intersections. Technically it could also replace the Batch Boolean algorithm, but that's not a good idea as it probably won't be as fast. 

In the paper this library's Boolean algorithm is based on, it makes clear that it only works for two input manifolds, not for removing the self-intersections of a single manifold. The reason is that the paper's method of linking edges can be done independently and still guarantee a manifold result, which is why the algorithm is parallelizable, but examples are given that break this guarantee in the single-manifold case.

However, the first half of the Boolean should still work basically the same. The collider will need to be updated so it can automatically skip intersections between a triangle and any edges that share a vertex with it. Then the calculation of intersection points and the winding numbers of vertices can proceed as it currently does.

The difference is there are a few new vertices that will not be found by these intersections of triangles with edges. These new verts will be the intersection of three non-neighboring triangles, which will be findable because they are also the intersection of three new edges, and this intersection exists if and only if the three new edges share a single set of three triangles as neighbors.

The bigger problem for manifoldness is that previously the new verts on a given partially-retained edge could be paired up arbitrarily (and independently) and still guarantee a manifold result. In the case of these new triple-intersections, that is no longer true - they depend on each other topologically. This means the assembly algorithm cannot be parallelized (that I know of), but by working serially it should be possible to start at an intersection and work our way along each seam, checking any previous neighbor pair-ups and basing the new pair-ups on those to maintain topological consistency. 

Additionally, another type of new edge can form, which has only one intersection vert. This is when a manifold is folded, producing an intersection curve that terminates on both ends instead of forming a loop. In the case when a vert opposite the intersected edge (the other vert of either neighboring triangle) is shared with the triangle it is intersecting, then this new edge will go from the intersection vert to this opposite vert.

**Comment by pca006132** [comment]
Apart from Warp, Refine can also cause overlap when the surfaces are too close.

**Comment by elalish** [comment]
The tricky part to making this robust is these new triplet vert (the name I'll give to verts at the intersection of three tris) pair-ups. In the Boolean version, we know for each vert on a retained edge whether it is a start or an end vert, and this pre-arranged consistency is fundamentally where our manifoldness guarantee stems from. In the case of these triplet verts, it is not known for each new edge whether they are start or end, only that within each partially-retained face, the two edges linking to that vert must be opposite (one start and one end), while on a given new edge there must be an equal number of start and end verts. 

This looks like a kind of binary integer programming problem, except all the constraints are equalities and there is no cost function - we only want an arbitrary valid solution. I'm not too concerned about NP-hardness, since in practice these knots of related constraints will be quite small and local. I'm guessing NP-hard means there is no clever way to solve this, so probably just need a little decision tree we can search exhaustively.

**Comment by pca006132** [comment]
No, the problem may just be NP hard in general, but exists efficient algorithms for special cases. For a lot of NP hard graph problems, we can solve them efficiently for planar graphs or tree like graphs. You can have a look at parameterized algorithms if you are interested. One important property is whether or not the problem can be split into subproblems locally, and it seems that it is possible.

**Comment by elalish** [comment]
Well, and actually on further reflection, I believe this is in fact an [XOR-SAT](https://en.wikipedia.org/wiki/Boolean_satisfiability_problem#XOR-satisfiability) problem, which is P because it can be solved by Gaussian elimination. Not sure if that's the route we want to take, but it's nice that this isn't as hard.

**Comment by elalish** [comment]
Okay, upon further reflection, the above approach isn't going to work. While an XOR-SAT might fix issues akin to figure 6.5 from Julien Smith's paper, I can't see a straightforward way to deal with the more common problems shown in figure 6.4.

I'm now inspired by his approach to robustifying the Bently-Ottmann algorithm for 2D overlap removal, from Section 7.4 and after. Here, instead of using shadow functions and symbolic perturbation to make every comparison into a reliable boolean value, now there are three options: `above`, `below`, or `on`. `on` is any comparison that is not definitely `above` or `below` (more like is a vert on one side of a line or the other - I'm not trying to imply coordinate direction). For a vert `on` and edge, that edge is split, so degeneracies are effectively snapped together. This has an extra benefit which is that it keeps degeneracies from entering the output in the first place (very much unlike the Boolean), which means we shouldn't have to spend as much time in our serial simplification post-process. 

I'm not a huge fan of the sweep line algorithm, especially since extending it to 3D sounds like an indexing nightmare and there's no opportunity for parallelism. However, I think with Julien's degenerate snapping, there's no need - the sweep line was a nice way to get from `O(n^2)` to `O(nlogn)`, but we can do that with our BVH tree anyway (and it's parallel). 

So, for the 2D problem (output an &epsilon;-valid polygon with e.g. a positive fill rule from an invalid polygon), it could look like this:

1. Merge all verts that are within &epsilon; of each other. &epsilon; is chosen as the maximum of a safe bound on inexact arithmetic error and the smallest feature size that is desired to retain. Care must be taken regarding chains of nearby verts so they don't move too far. They need to use a weighted-average position and calculate distance based on up-to-date values. The broad phase can be parallel, but the narrow phase probably needs to be serial for this reason.
2. Discard collapsed edges - the remaining set are the "original" edges. 
3. Every edge builds an ordered list of all verts that are within &epsilon; of it. The order is reliable even with inexact arithmetic because any verts that were near each other have already been collapsed. 
4. Find new edge-edge intersections - they cannot include any where an end-vert of one edge is in the vert list of the other. When sorting the new vert into each edge's list, if it is within &epsilon; of its neighbor, it is not inserted and that neighbor vert is inserted instead into the other edge's list. There can be no nearly-parallel cases, and so the cyclic ordering of the edges around an intersection vertex is always clear, even with inexact arithmetic. 
5. Break the edges into sub-edges according to their internal sorted vert list. None of these sub-edges intersect each other.
6. Output only the set of edges that separate winding numbers 0 and 1. This can be done in parallel by ray-casting vertically or horizontally from the midpoint of each edge (to avoid the ray being parallel to its edge) to evaluate the winding number. This is robust by just using symbolic perturbation to check edge shadowing, since no verts or edges are close to the midpoint vert.

For the 3D case, things get a bit more complicated, but follow the same general concept. Before, after step 1, verts went from having exactly 2 edges to having an even number of edges attached. Now edges will go from 2-manifold to even-manifold (number of triangles they're attached to). And instead of outputting sub-edges that divide winding numbers 0 and 1, we need to output sub-polygons of input triangles that divide winding numbers 0 and 1. These polygons will be triangulated as a post process, just like in the Boolean. Steps:

1. Same as 1 above.
2. Same as 2 above, though triangles also get collapsed and discarded naturally. Edges with the same two verts are merged and create a list of their attached triangles (an even number).
3. Same as 3 above.
4. Same as 4 above - this is now another degenerate case, as edges would all be skew with input in general position.
5. Every triangle builds a list of verts that are within &epsilon; of its interior. None need to be merged.
6. Find new edge-tri intersections - they cannot include any pairs already covered by steps 4 or 5. Insert the new vert into the edge and tri lists, unless it is close to one of the existing verts in the list, in which case it is merged instead, as in 4. 
7. Add a new edge for each tri-tri intersection. The endpoints will be the exactly two verts that are shared between the edges of one tri and the edges and interior of the other tri and vice-versa. I believe this will always be true based on the merging we've done, but a proof would be nice.
8. Add to this new edge vert list any verts from the interior of either triangle that are within &epsilon; of it. 
9. Find new intersection verts between new edges within each tri. Insert or merge the new vert into both edges' lists as in 4.
10. Same as 2D step 5, but now each triangle also gets a list of halfedges and the sub-edges are pushed as half-edges into each of their triangles's lists (the new edges insert both a forward and backward halfedge into each triangle) .
11. Each triangle now has a set of halfedges that looks like the output of 2D step 5. Join them into a valid polygon partition by  following the cyclic ordering around each vert, which is unambiguous since no sub-edges are near parallel.
12. Output only the polygons that separate winding numbers 0 and 1. The polygons may not be convex, so a midpoint can't work for ray-casting. We could triangulate at this stage and use the midpoint of one of the interior edges, though it would mean doing twice as much triangulation as we need to. We could also ray-cast from a vert and handle the other faces that share it by checking face shadowing. That might be a better approach in 2D as well.

This feels implementable - more serial than the Boolean, but still parallel where it really matters (the BVH broad phase). Curious if anyone can think of an example case where this could break down?

**Comment by zmerlynn** [comment]
Following up on [your sketch](https://github.com/elalish/manifold/issues/289#issuecomment-2111069955): I took a stab at building out the 2D case alone, partly to see if the approach survives contact with implementation, partly to give us something concrete to discuss.

Branch: https://github.com/zmerlynn/manifold/tree/explore/overlap-removal-2d
- Design write-up: https://github.com/zmerlynn/manifold/blob/explore/overlap-removal-2d/docs/Overlap2D.md
- Code: https://github.com/zmerlynn/manifold/blob/explore/overlap-removal-2d/extras/overlap2d_proto.cpp

Quick disclosure: most of the implementation was driven by Claude Code, with me pushing it to research and resolve weaknesses I noticed, and reviewing the output. (I'm sure this is not a surprise given my other PRs.)

The shape follows your sketch (BVH pair queries, the 6-step pipeline, reuse of the 3D building blocks). For parts the sketch didn't fully pin down, I leaned on Smith §7, mostly for the correctness framework (eps formula, iteration convergence proof, symbolic perturbation predicates) rather than its sweep-line infrastructure. The three local divergences from both are documented in the design doc, each with what prompted it.

I'm not assuming you want this in any particular form. The design doc proposes landing the prototype in extras/ as reference, but "interesting, not the direction we want" or "let's discuss inline here without merging anything" are equally reasonable answers. Happy to take this whichever way is most useful to you.

**Comment by elalish** [comment]
I'm super excited to see this, and it looks like excellent progress! I'll respond to your open questions here for lack of a better place.

>Landing site in src/. New src/overlap2d/, or wired into cross_section.cpp directly as the boolean/Simplify backend. Driven by how much shared code with 3D booleans we expose.

Directly wired into cross_section.cpp sounds good to me. The sooner we can remove the Clippper2 dependency, the better, especially in light of #1683, which I could work around, but I frankly find annoying. Pretty much the only other thing we use it for is offset, and that might not be too hard if we just push the verts along their bisectors and then fix all the self intersections with what you have here.

>eps policy. Boolean2D accepts explicit eps or auto-infers when ≤0. CrossSection's existing API doesn't expose eps; add it as an optional parameter, or always auto-infer? Smith's α-budget is conservative (kbudget=1000); Simplify may want stricter.

I've intended for some time to make the CrossSection API match Manifold better by including user-definable tolerance. This may be the right time to introduce it. 

>Multi-threading. manifold::Collider is parallel-ready; the prototype is MANIFOLD_PAR=-1 only. Step 1 union-find and step 6 per-face ray-cast are the natural parallelism candidates.

👍 

>Additional quantitative oracles. The prototype now has area preservation and area-drift tracking on top of the topology check (see Validation), and consumes polygon_corpus.txt directly. The remaining easy add is Clipper2's Polygons.txt (195 numbered cases with SOL_AREA references), which gives an independent third-party oracle on a wider input distribution than polygon_corpus.txt. Worth doing for production landing; not strictly required for the design discussion.

Agreed, pulling in their tests sounds great. Does their license allow it?

At the scale of what you and Claude have written, there's no way we'll be able to judge correctness by code review. We can give style pointers and such, but we'll need to lean heavily on testing to convince ourselves this is exceeding the standard of Clipper2 (which should be doable, especially considering the precision issue we already have). As you do get this ready as an MR for review, it would be great to show two metrics: performance vs our existing for polygon ops, and the delta in our WASM build size. I'm not going to gate based on these, as there are lots of factors, but I think it's worth keeping them in mind.

**Comment by hrgdavor** [comment]
Does this lead to manifold working without clipper2 for 2d booleans and stuff ? I am excited to see where this goes :)

**Comment by zmerlynn** [comment]
Quick perf + direction question.

Perf head-to-head vs `Clipper64` (per #1683's direction toward `Paths64`): 6x slower on mfogel (62 cases), ~10x on JTS overlay (2042 large adversarial), ~46x on Clipper2's own corpus (21 small/typical polygons). No real perf tuning yet, and the full-FP design is the load-bearing bottleneck - standard hardening probably gets to ~5x, not parity.

Also worth flagging: I tested empirically and Clipper2 handles every overlap pattern the prototype targets cleanly via `UNION`+`NONZERO` - bowtie self-crossings, nested overlapping triangles, heavy-self-crossing star polygons, pinched/self-touching, mixed-orientation rings. The prototype isn't filling a correctness gap there.

Re-reading the thread, the actual motivation for 2D looks like it was using it as a stepping stone for robustifying the 3D Boolean against non-manifold input from `Warp` / `Refine` - your line about the 3D case *"following the same general concept"* was the load-bearing rationale, not Clipper2 replacement. I'm prototyping the 3D side along the way too, in a sibling branch, sharing the Smith framework - though it'll be a bit before it's ready to show.

Two paths fall out, with different priorities:

- **3D framework primary**: stay FP in 2D, accept some perf hit, keep 2D and 3D structurally aligned with the Smith machinery.
- **Dep-removal primary**: integer rewrite of the same 6-step architecture (BVH / DCEL / winding-filter, exact int128 kernel), then offset on top. Likely closes most of the perf gap. Decouples from the 3D framework - 2D and 3D become cousins rather than siblings. WASM only actually shrinks once both booleans *and* `InflatePaths` are migrated.

Which goal matters more to you? Path follows from that.

**Comment by elalish** [comment]
Yes, this is intended to be a stepping stone to 3D, but I'd also like better codebase consistency with 2D and to remove a dependency. I tend to think of 2D problems as trivial compared to 3D problems time-wise (I'm sure that's not always true, but I think in practice it's a good approximation), so I'm not too concerned about even a 6x difference. And these timings are for single threading too, right? So ideally we should get some speed from parallelization; I don't think Clipper is parallelized. 

I'm looking for consistency, so let's stick with FP across the board - no need to switch to integer for a bit of performance. Once upon a time I considered writing this whole library in fixed precision, but I'm glad I went with the decision I did. The simplicity is worth it, even if there's some speed penalty. 

How do you feel about migrating InflatePaths? If it's not too much of a detour, I think it would be nice to button up 2D before moving on to 3D.

**Comment by zmerlynn** [comment]
Continuing the 3D thread from [your comment on #1707](https://github.com/elalish/manifold/issues/1707#issuecomment-4899722953) - leaving that PR for the 2D backend.

Honestly, I went down a bit of a blind alley chasing exact arithmetic on the 3D side, and two course-corrections both landed me back on your #289 sketch: coming back to SoS returned me to the existing kernels, and dropping the sweep for a batch arrangement is essentially what #289 describes anyway.

So: the beginnings of a prototype following that 13-step sketch, staying full-FP / inexact SoS. Still early - I'll share more as it firms up.

On the public API, a couple of things worth settling:

Name - I'm leaning `RemoveOverlaps` (matches this issue, reads clearly); I considered `Regularize` too.

Semantics:
- Positive winding as the fill rule, same as 2D - overlapping/nested coverage collapses to one solid, and inverted (negative-winding) regions come out empty.
- The output is a single Manifold but may be multi-component: a self-overlapping input can split into separate pieces - two spheres touching only at a point, or cubes joined only along an edge, come out as distinct closed solids rather than a non-manifold fusion at the shared feature (Decompose to separate). Worth pinning down since it shapes the contract.

**Comment by elalish** [comment]
Thanks for the update - what does SoS stand for? Out of curiosity, how should we rationalize dropping the sweep for batch in 3D when we came to the opposite conclusion in 2D? Granted, if we can make batch work I'd be thrilled, but if it's blocked in 2D, I'm sure 3D will have at least as many problems. Did we gain insight that will also help 2D?

I like `RemoveOverlaps`. Positive fill is good. Output as multicomponent makes perfect sense. Also, the Boolean generates 4-manifold edges occasionally, so part of our simplifier fixes those up by duplicating verts to get back to 2-manifold. You may be able to take the same path, if that makes it easier.

**Comment by zmerlynn** [comment]
SoS is Simulation of Simplicity (Edelsbrunner & Mucke) - the symbolic-perturbation tie-break for degenerate cases. Smith's `+xi` is that same idea carried into the inexact setting.

On sweep vs batch: I think - but am not sure yet - that I've cracked how to do it in batch, and I'm still working through it. The ingredient is a global whole-block resolution, the same thing the 2D sweep relies on, applied over a materialized arrangement; it looks like the piece the original 2D batch sketch was missing, so if it holds it should carry back to 2D. You're right that it's the same problem in both dimensions - I don't think 3D is easier here - so I'm checking it against an independent oracle before I claim anything.

RemoveOverlaps / positive fill / multicomponent all sound right to me. And the 4-manifold-edge tip is helpful - reusing the simplifier's vert-duplication path for the manifold output sounds easier than reinventing it.

**Comment by elalish** [comment]
Excellent, I look forward to seeing how it works. Might be easier to review the new 2D version before opening a 3D PR, but let me know what works best for you.

**Comment by hassansaleh31** [comment]
I was about to file an issue but I stumbled upon this one, I'm trying to render a wire form that bends on itself for fishing lures, it's for an app I'm building for lure makers and designers, here's an example I found on google:

<img width="448" height="272" alt="Image" src="https://github.com/user-attachments/assets/fafc7679-377b-437c-905c-840fc32af476" />

Generating the sweep from 2D rings works but contained lots of self-crossings

<img width="874" height="381" alt="Image" src="https://github.com/user-attachments/assets/d8bc6250-65fa-448f-98d2-ab0fc45751cc" />

I did manage to solve it eventually by splitting the form into 11 sections and joining them with a union operation which got rid of the overlapping vertices. But since these forms are user generated, I will have to somehow figure out a way to dynamically split each form.

A sweep feature from manifold would be amazing, or a `fill` parameter similar to how it is done in 2D.

**Comment by zmerlynn** [comment]
JFYI I haven't abandoned this, but it's been a series of grinding through one degenerate case after another so I've been kind of dark. I will update more soon!
